### PR TITLE
22 terraform ecs

### PR DESCRIPTION
**Pull request by oskarwilliams378**
#22 
Add terraform to deploy:
ECS
LB
CloudFront
WAF

and subsequent dependencies

### CODE REVIEW

File: terraform/demo/main.tf [hunk 1]
```diff
@@ -48,3 +62,16 @@ module "lev_api" {
   }
   environment_name = "demo"
 }
+
+
+module "data-share-service" {
+  source = "../modules/data-share-service"
+  providers = {
+    aws.us-east-1 = aws.us-east-1
```
Comment (olevett) [inline terraform/demo/main.tf:70]: We presumably need this for CloudFront?

File: terraform/demo/main.tf [hunk 2]
```diff
@@ -48,3 +62,16 @@ module "lev_api" {
...
+  }
+  environment                 = "demo"
+  ecr_url                     = "${data.aws_caller_identity.current.account_id}.dkr.ecr.eu-west-2.amazonaws.com"
+  service_port                = 8080
+  cloudwatch_retention_period = 30
+  vpc_cidr                    = "10.158.16.0/20"
```
Comment (olevett) [inline terraform/demo/main.tf:76]: is this just a random choice?

File: terraform/demo/main.tf [hunk 3]
```diff
@@ -48,3 +62,16 @@ module "lev_api" {
...
 }
+
+
+module "data-share-service" {
+  source = "../modules/data-share-service"
+  providers = {
+    aws.us-east-1 = aws.us-east-1
+  }
+  environment                 = "demo"
+  ecr_url                     = "${data.aws_caller_identity.current.account_id}.dkr.ecr.eu-west-2.amazonaws.com"
```
Comment (olevett) [inline terraform/demo/main.tf:73]: I _think_ I'd like this to come out of the ecr module as an output, but lets not worry about it for now (it'll make the non-demo environment trickier anyway)

File: terraform/demo/.terraform.lock.hcl [hunk 4]
```diff
@@ -1,6 +1,24 @@
...
+    "zh:841fc4ac6dc3479981330974d44ad2341deada8a5ff9e3b1b4510702dfbdbed9",
+    "zh:91be62c9b41edb137f7f835491183628d484e9d6efa82fcb75cfa538c92791c5",
+    "zh:acd5f442bd88d67eb948b18dc2ed421c6c3faee62d3a12200e442bfff0aa7d8b",
+    "zh:ad5720da5524641ad718a565694821be5f61f68f1c3c5d2cfa24426b8e774bef",
+    "zh:e63f12ea938520b3f83634fc29da28d92eed5cfbc5cc8ca08281a6a9c36cca65",
+    "zh:f6542918faa115df46474a36aabb4c3899650bea036b5f8a5e296be6f8f25767",
+  ]
+}
+
 provider "registry.terraform.io/hashicorp/aws" {
```
Comment (olevett) [inline terraform/demo/.terraform.lock.hcl:22]: Is there a reason we don't have the random provider in a lock here?

File: terraform/modules/data-share-service/cloudfront.tf [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      # AWS enforce a maximum of 60s, but we can request more if desired.
+      # See https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/distribution-web-values-specify.html#DownloadDistValuesOriginResponseTimeout
+      origin_read_timeout = 60
+    }
+  }
+
+  enabled         = true
+  is_ipv6_enabled = false
+
+  #  logging_config {
```
Comment (olevett) [inline terraform/modules/data-share-service/cloudfront.tf:23]: Do we have a ticket/something tracking this logging set up?

File: terraform/modules/data-share-service/ecs-security.tf [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  vpc_id      = module.vpc.vpc_id
+
+  ingress {
+    protocol        = "tcp"
+    from_port       = var.service_port
+    to_port         = var.service_port
+    security_groups = aws_security_group.lb_auto.*.id
+  }
+
+  egress {
```
Comment (olevett) [inline terraform/modules/data-share-service/ecs-security.tf:13]: This is a broad "allow" right? I think that's fine for POC, but we should lock this down as and when things have stabilised and we know what legit outbound traffic is.

File: terraform/modules/data-share-service/ecs.tf [hunk 7]
```diff
@@ -0,0 +1,60 @@
+resource "aws_ecs_cluster" "main" {
+  name = var.environment
+}
+
+resource "aws_ecs_task_definition" "gdx_data_share_poc" {
+  family                   = "${var.environment}-gdx-data-share-poc"
+  requires_compatibilities = ["FARGATE"]
+  network_mode             = "awsvpc"
+  memory                   = 512
+  cpu                      = 256
```
Comment (olevett) [inline terraform/modules/data-share-service/ecs.tf:10]: This seems fairly low for a JVM service, but I guess we'll have to wait and see

File: terraform/demo/main.tf (hunk 1 above)
Comment (oskarwilliams378) [reply terraform/demo/main.tf:70]: Yep

File: terraform/demo/main.tf (hunk 3 above)
Comment (oskarwilliams378) [reply terraform/demo/main.tf:73]: Yeah tracking that across to non demo is an issue.

Also all it is is the account number followed by a constant string, so it is less important to output

File: terraform/demo/.terraform.lock.hcl (hunk 4 above)
Comment (oskarwilliams378) [reply terraform/demo/.terraform.lock.hcl:22]: I haven't init'ed there, done

File: terraform/demo/main.tf (hunk 3 above)
Comment (olevett) [reply terraform/demo/main.tf:72]: I think we should be able to do some remote state/similar, but lets not worry for now

File: terraform/modules/data-share-service/cloudfront.tf (hunk 5 above)
Comment (oskarwilliams378) [reply terraform/modules/data-share-service/cloudfront.tf]: Added logging now

File: terraform/modules/data-share-service/ecs-security.tf (hunk 6 above)
Comment (oskarwilliams378) [reply terraform/modules/data-share-service/ecs-security.tf:15]: Other services I have seen have been perfectly happy with this, and it is the default with AWS, but am happy to lock down later

File: terraform/modules/data-share-service/ecs.tf (hunk 7 above)
Comment (oskarwilliams378) [reply terraform/modules/data-share-service/ecs.tf:15]: Yeah I can imagine so, I just left this as a start point

File: terraform/demo/main.tf [hunk 8]
```diff
@@ -48,3 +62,16 @@ module "lev_api" {
   }
   environment_name = "demo"
```
[the remark this answers is not in the document]
Comment (oskarwilliams378) [reply terraform/demo/main.tf:63]: Done

**Review by olevett — APPROVED** [review]

**Pull request merged**